src/RecursiveStagesService.php

```diff
@@ -0,0 +1,187 @@
+<?php
+
+namespace SilverStripe\Versioned;
+
+use Exception;
+use Psr\Container\NotFoundExceptionInterface;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Core\Resettable;
+use SilverStripe\ORM\DataObject;
+
+/**
+ * Functionality for detecting the need of publishing nested objects owned by common parent / ancestor object
+ */
+class RecursiveStagesService implements RecursiveStagesInterface, Resettable
+{
+    use Injectable;
+
+    private array $stagesDifferCache = [];
+    private array $ownedObjectsCache = [];
+
+    public function flushCachedData(): void
+    {
+        $this->stagesDifferCache = [];
+        $this->ownedObjectsCache = [];
+    }
+
+    /**
+     * @return void
+     * @throws NotFoundExceptionInterface
+     */
+    public static function reset(): void
+    {
+        /** @var RecursiveStagesInterface $service */
+        $service = Injector::inst()->get(RecursiveStagesInterface::class);
+
+        if (!$service instanceof RecursiveStagesService) {
+            // This covers the case where the service is overridden
+            return;
+        }
+
+        $service->flushCachedData();
+    }
+
+    /**
+     * Determine if content differs on stages including nested objects
+     * This method also supports non-versioned models to allow traversal of hierarchy
+     * which includes both versioned and non-versioned models
+     *
+     * @param DataObject $object
+     * @return bool
+     * @throws Exception
+     */
+    public function stagesDifferRecursive(DataObject $object): bool
+    {
```

Comment:
```suggestion
    public function stagesDifferRecursive(?DataObject $object = null): bool
    {
        if (!$object) {
            $object = $this->owner;
        }
```
You'll need to make this adjustment when moving this into `Versioned` so we don't have to call it like `$obj->stagesDifferRecursive($obj)`

Reply:
This is pending the overall discussion about the code downgrade.